### ISSUE TITLE
uses wayyy to much RAM freezing the machine. 

### ISSUE THREAD

**Issue by nojvek** [codota/TabNine#182]
I'm on a 16GB Macbook Pro. I occasionaly see TabNine chew up like 4GB of memory. It brings the machine to a dead stop as OS has to swap things in and out.

Let's make TabNine a bit more efficient

**Comment by andi1984** [comment]
I really love the idea and the VSCode plugin. It amazed me in which kind of situations it helped me, but I also have to say that on especially large code projects it is getting really hairy so that I also had to disable the plugin in those workspaces. Any small improvement on performance would be already nice.

**Comment by Shougo** [comment]
You can disable Tabnine local by Tabnine::config.

**Comment by jaydorsey** [comment]
Also on a 16GB Macbook Pro. CPU usage is routinely 1.5+GB usage, even at idle; fans spinning _all_ the time.

Once I kill TabNine, disable Deep TabNine Local, and restart vim, it starts at 70MB and grows until about 450MB.

**Comment by amircodota** [comment]
Hi, everyone who wrote on this issue.

We have since released a version that should improve TabNine's cpu and memory consumption.
Would really appreciate your help in trying it out and seeing if the situation improved.

Thanks a lot!

**Comment by jaydorsey** [comment]
I'm using 2.5.1w/ coc-tabnine. I'm still showing 2.10GB of RAM usage with a fresh vim instance, editing just 1 or 2 lines of code

This is while editing a ruby file. TabNine local is enabled, everything else is disabled (I also have solargraph enabled so I don't know if that's related). It's on a fairly large repository, not sure if that matters

Disabling TabNine local, I see similar behavior (in contrast to the past, when it showed reduced memory consumption)

**Comment by amircodota** [comment]
@jaydorsey Thank you for that valuable feedback.

Regarding RAM usage - I'm guessing it is a pretty big project. That memory usage can be seen for a big project. The difference from before is that it should not keep growing, where before it would grow and grow, especially on linux machines.
Regarding cpu usage - we have made some improvements. We still have some fixes we want to try - hopefully it will alleviate the pain of fans spinning. I'm sure you understand that TabNine local is a pretty heavy thing, so picks of high cpu are to be expected, but it should not use cpu constantly - definitely not while you are idle.

Regarding not making a difference by switching TabNine local off - this sounds like a bug. Before when tabnine would detect a config change, it would restart itself. I guess it did not happen for you. I will look into it.

Thanks again for that valuable feedback.

If anyone else on this thread would like to test the version and let us know if things improved - it would be great.

**Comment by jaydorsey** [comment]
@amircodota I've been monitoring it over the last few days and it does appear to be relatively stable, with the size varying only whether TabNine Local is enabled or not. I'm not 100% sure about the restart after switching (not sure how to test) but I did notice (maybe related) that magic strings aren't working (using nvim/coc-vim) after I use `TabNine::config`. I have to shut nvim down completely and re-open before any other magic string works again. No error messages

**Comment by amircodota** [comment]
@jaydorsey glad to hear things are improving in terms of performance.

Regarding magic strings - just to make sure I understand - the first time you type `tabnine::config` it works, but then any magic string (tabnine::config, tabnine::sem etc.) will not work until you restart nvim?

**Comment by jaydorsey** [comment]
> Regarding magic strings - just to make sure I understand - the first time you type `tabnine::config` it works, but then any magic string (tabnine::config, tabnine::sem etc.) will not work until you restart nvim?

That was the behavior I thought I was seeing, but I've narrowed it down a little further and it's not quite that bad. It looks like the magic strings only work once per buffer (which is fine). If I clear out my buffer and remove any previous reference to the magic string I'm typing, the magic strings work as expected

When I tested it before it appears I may have been typing it on screen multiple times, on new lines, in the same buffer, without deleting previous occurrences. I would open config with `TabNine::config`, do a change to it, close the browser tab, restart, and try to re-open the config:

```
TabNine::config
TabNine::restart
TabNine::config # this one wouldn't work
```

Works fine to just clear the buffer out after each magic string

tldr; Operator error/misunderstanding

**Comment by amircodota** [comment]
@jaydorsey 
I'm afraid I do not know vim well enough to understand is this is severe or not. Do you think other How likely other people will hit this and be confused by it?

We hardly touched anything in that area, so I doubt this is a regression, though.

**Comment by jaydorsey** [comment]
I wouldn't consider it severe, and I don't think anyones likely to run into it. 

The only reason I even noticed it is because I was doing testing to try and reproduce behavior with the RAM usage so visiting & revisiting the configs and such and being lazy with using the same buffer for the text. Average user isn't going do what I did. I've noticed this behavior for a while and just restarted neovim each time; didn't even make the connection to try and debug root cause until you'd mentioned above I can do restart and it should work

**Comment by amircodota** [comment]
opened the [issue under tabnine-vim](https://github.com/codota/tabnine-vim/issues/54)

Closing thisissue as it originally refers to memory issue

**Comment by erfanium** [comment]
![Screenshot from 2020-05-04 03-12-09](https://user-images.githubusercontent.com/47688578/80927744-26562c00-8db5-11ea-9ab6-8cca3578f6f7.png)

**Comment by amircodota** [comment]
@erfanium 1.5GB is expected if you have deep tabnine local enabled.

You can choose to disable deep tabnine local.
In addition, you can register for the pro version (paid after trial), in which case you can use deep tabnine cloud instead of deep tabnine local.
